### PR TITLE
New dialog component

### PR DESCRIPTION
**Pull request by stevedya**
Addresses #8475. This PR is for the creation of a completely new and versatile dialog component. 
Screen Shot 2022-05-10 at 11 58 23 AM

Using dialogs in admin templates can new be generated with 2 template tags:

### The toggle
```
 {% dialog_toggle class_name='button button-primary' dialog_title='Modal title' text='Open simple dialog' %}
```
### The dialog 
```
{% dialog icon_name="globe" title="Modal title" description="This is a description" message_status="info" message_heading="Info message title" message_description="This is a subtext" %}
   <!-- content here -->
{% enddialog %}
```

### Dialog message types
There are 4 types of messages a dialog can have: `info | success | critical | warning `

**Other screenshots here:** 
[Dialogs.zip](https://github.com/wagtail/wagtail/files/8663896/Dialogs.zip)

### Browsers tested on 
Chrome 100, Safari 15.2, Firefox 99
Windows: chrome 99

### CODE REVIEW

File: client/src/tokens/colors.js
Comment (stevedya) [inline client/src/tokens/colors.js:162]: After design review with Ben we decided to modify the hue of the warning color. Currently this isn't used anywhere either.

File: wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html
Comment (lb-) [inline wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html:1]: it is worth setting up a template var or 'with' for the id, so save having to write it multiple times

`{% with id=title|cautious_slugify %} `

or a template tag?

`{% get_id from=title as id %}`

File: wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html
Comment (lb-) [inline wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html:7]: will there be any way to add custom classes to the dialog?

File: wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html
Comment (lb-) [inline wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html:30]: is it worth adding a subtitle section also? as a span inside the h2.

File: wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html
Comment (lb-) [inline wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html:36]: it might be worth putting a comment that the 'end' is intentionally missing and where to look for it.

File: wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html
Comment (lb-) [reply wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html:7]: Should this be a `template` tag instead? and when the modal opens it will put the template contents at the end of the DOM, otherwise there is a chance these dialogs will be added anywhere and adopt the styling of their container.

Comment (lb-) [reply wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html:7]: Will there be any way to add additional data attributes here (or attributes in general)?

File: wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html
Comment (lb-) [inline wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html:34]: Sorry @stevedya I think that this should stay as description. 

However I think we could also add a subtitle that appears within the title .

**Comment by lb-** [comment]
A few other questions
* Will we be documenting the events for this anywhere? if so, will we maybe re-emit the events with our preferred naming prefix? https://a11y-dialog.netlify.app/advanced/events
* Will we need any kind of scroll lock? https://a11y-dialog.netlify.app/advanced/scroll-lock/
* Will we offer a way to disable closing the modal by clicking outside (maybe as a future enhancement)
* If the dialog is a message dialog - will we define it as such using `alertdialog` https://a11y-dialog.netlify.app/advanced/alert-dialog  (kind of relates to the above)
* Will we provide more template blocks in `wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html` so that anyone extending this template can add content inside as needed.

File: wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html
Comment (stevedya) [reply wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html:1]: Great idea! I've went with the `{% with %}` option

File: wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html
Comment (stevedya) [reply wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html]: - **custom classes / attrs:** I spoke with @thibaudcolas about this but he wasn't keen on having these as an option at this point. I do think having abilities to pass custom classes is definitely useful, although i'm not sure how much customization we are aiming to allow on this component as well. For now i've made an option for `{{ class_name }}` at least and will take handle any lashings from Thibaud 😁   

- **template tag** - Awesome suggestion/improvement! I've made this change now 😁

File: wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html
Comment (stevedya) [reply wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html:12]: Yep for sure! The comment on line 35 before was intended for that originally but definitely wasn't very clear haha 😅 I've tried to clarify it a little more now, and moved it closer to the bottom.

File: wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html
Comment (stevedya) [reply wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html]: I'm not so sure on adding something like that to the h2 because it's intent is to be used as the title. I'm also not sure what value this would provide, because if someone would like to add extra text to the end of the title they could append it to the string when passing `title` or use the provided `subtitle` attr. (Also trying to keep the template tag as lean as possible.)

I do think renaming description to subtitle was a welcomed changed because it is more inline to how the designs appear:
Screen Shot 2022-05-10 at 11 50 17 PM

Comment (lb-) [reply wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html]: good point  - maybe skip the subtitle idea. 

I think description (what you had before) is more suitable for a Dom element outside the header element. 

This is because we kind of have this convention already in the page headers subtitle appears inside the h1 and description (when used) is outside the h1 in its own element. 

It also gives us the ability to add subtitle in the future if the need arises.

File: wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html
Comment (lb-) [reply wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html:12]: Thanks. Sorry i didn't see it before. The updated one is heaps clearer now.

File: wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html
Comment (lb-) [reply wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html]: Great. Let's do attrs later if needed. 

I think classname is a minimum requirement for these things in my opinion. Easy to add and gives a lot of flexibility for custom usage. Plus low risk in terms of html injection issues.

**Comment by stevedya** [comment]
> A few other questions
> 
> * _Will we be documenting the events for this anywhere? if so, will we maybe re-emit the events with our preferred naming prefix?_ 
**Great question, I think this could use a UI team discussion (but I don't know if we should let it hold up the base setup work). In the meantime, I could add a few lines of documentation to the JS file that includes a sentence about events with a link to their docs for now?**

> * _Will we need any kind of scroll lock?_ 
**It had seemed to be working by default but maybe the browser was being smart. Anyway, I've added this now to be on the safe side.** 
> * _Will we offer a way to disable closing the modal by clicking outside?_ 
**This is enabled by default** 🎉 🥳 
> * _If the dialog is a message dialog - will we define it as such using `alertdialog` https://a11y-dialog.netlify.app/advanced/alert-dialog  (kind of relates to the above)?_ 
**For now I believe all the dialogs are intended for components with user interaction and not alerts so this might be a future enhancement** 
> * _Will we provide more template blocks in `wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html` so that anyone extending this template can add content inside as needed._ 
**Great question, however I believe we would persuade developers to utilize the tag with a nested children of their customizations, instead of extending. (I think extending it would cause some breakage and need special consideration because of the unclosed div tags as well)** 
 eg ```{% dialog %}<p>Your dialog stuff here</p>{% enddialog %}``` 
 (right now that is equivalent to populating the `{% block content %}` of the dialog)

**Comment by lb-** [comment]
Thanks for the response @stevedya those make sense. I might make a few more comments for the template block as it may have not been clear.

File: wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html
Comment (lb-) [inline wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html:16]:
```suggestion
                {% block closebutton %}
```

If we used blocks for these, including subtitle message and others we solve two two problems. 1 - we can define the section instead of comments and 2 - when using template inheritance these blocks can be overridden or extended. 

Note that template overrides are a documented way to do some kinds of customisations and it seems reasonable to provide these here.

File: wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html
Comment (lb-) [inline wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html:19]:
```suggestion
                </button>
                {% endblock%}
```

**Review by kaedroho — CHANGES_REQUESTED** [review]
As we are past the 3.0 release, I think we need to talk about Tailwind before accepting more changes that use it significantly.

This PR and some of your previous PRs use Tailwind instead of SCSS for styling.

Our current [code style guidelines](https://docs.wagtail.org/en/stable/contributing/css_guidelines.html) make no mention of Tailwind and there is no RFC or GitHub discussion about it either.

I think we need to discuss this as I am no clear on what the overall plan is, and when we should and shouldn't use Tailwind. I'm sure some other core team members are wondering this too.

File: wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html
Comment (stevedya) [reply wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html]: Agreed!

File: wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html
Comment (stevedya) [reply wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html]: Yep! I've fixed this now :)

File: wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html
Comment (stevedya) [reply wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html:16]: That makes sense and I'm totally happy to add those in for sure, but as far as I know this component was intended to be used as a template tag itself (not as an include or extended from). 
I'm also concerned if someone where to extend this template, it would be broken because of the unclosed `div && template` tags

Comment (lb-) [reply wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html:16]: fair enough, I think for the scope of this it makes sense to leave that use case out for now.

**Review by thibaudcolas — CHANGES_REQUESTED** [review]
@stevedya looking pretty close. Let’s re-implement this with less Tailwind, more BEM, so we can unblock this.

In particular I’d expect everything inside `dialog-message.html` and `dialog.html` to use BEM with a new `components/_dialog.scss` declaring a `.w-dialog` block and its elements & their modifiers.

1. Make a PR refactoring `@apply` so we can compare
2. Then make a PR with vanilla CSS properties and values with `theme(`

File: wagtail/admin/templates/wagtailadmin/shared/dialog/dialog-message.html
Comment (stevedya) [reply wagtail/admin/templates/wagtailadmin/shared/dialog/dialog-message.html:14]: I've brought this back into the main dialog now, great point!

File: wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html
Comment (lb-) [inline wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html:34]: Double checking - is this going to change back to description?

Comment (stevedya) [reply wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html:38]: I'm quite keen on keeping it as a subtitle or perhaps we could rename this and the title above to: `heading / subheading` because thats what it more accurately represents in the visual styling of this component.  
### screenshot for visual example 
Screen Shot 2022-05-13 at 4 06 13 PM
 
I'm a little lost on what value it would add to have: `<h2>{{ title }}{{ subtitle}}</h2>` vs `<h2>{{ title }}</h2>` in the future, since they would render the same by passing your entire string (title & subtitle) to {{ title }}. 

Unless you're suggesting that title && subtitle both reside inside of the h2 but are styled similar to the screenshot above so that it all registers as 1 heading to assistive tech?

Sorry for the confusion too

Comment (lb-) [reply wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html]: My reasoning is the previous convention set by the header template. 

See this PR for context. 
https://github.com/wagtail/wagtail/pull/8368

- if we have a subtitle - I think it should reside within the h2 in a span, it is part of the title just visually shown different
- we should keep the original description content and have this sit in a div outside of the header. P tag is also good

Comment (lb-) [reply wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html]: Happy with heading/subheading but just trying to avoid too much different naming of things across similar (conceptually) content. 

But I understand that the modal is not a page and hence the title/subtitle could be reserved for that only. 

I know these are nitpicky but just trying to look at the conventions we have set elsewhere and try to honour them in new code (where it makes sense).

Comment (stevedya) [reply wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html]: > * if we have a subtitle - I think it should reside within the h2 in a span, it is part of the title just visually shown different


Awesome! Thanks for the clarification. I have moved this to reside inside of the h2 to bring it more in sync with the other title/subtitle combos we have.

**Comment by stevedya** [comment]
Hey @thibaudcolas 
I have made these changes now: 
 
> In particular I’d expect everything inside `dialog-message.html` and `dialog.html` to use BEM with a new `components/_dialog.scss` declaring a `.w-dialog` block and its elements & their modifiers.
> 
> 1. Make a PR refactoring to `@apply` so we can compare

I have made this change this inside this PR now :) 

> 2. Then make another PR with vanilla CSS properties and values with `theme(` functions so we still pull in our design tokens

Here is the other PR #8541  branched from this one with the css converted to using css props/theme() function (no @apply).

File: wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html
Comment (lb-) [reply wagtail/admin/templates/wagtailadmin/shared/dialog/dialog.html]: Thank you. This is much cleaner.

**Comment by thibaudcolas** [comment]
Closing in favour of #8541 after UI team discussion.

**Pull request closed without merging**
